### ISSUE TITLE
SVGElement.getBoundingClientRect(): which bounding box?

### ISSUE THREAD

**Issue by progers** [w3c/svgwg#339]
The CSSOM spec states (https://drafts.csswg.org/cssom-view/#dom-element-getclientrects):

> If the element has an associated SVG layout box return a sequence containing a single DOMRect object that describes the bounding box of the element as defined by the SVG specification, applying the transforms that apply to the element and its ancestors.

The SVG spec now has a definition of the bounding box but there are three versions: object, stroke, and decorated (https://svgwg.org/svg2-draft/single-page.html#coords-BoundingBoxes).

This has been discussed before:
https://lists.w3.org/Archives/Public/www-svg/2011Jul/0054.html

Testcase: http://output.jsbin.com/fuxuqav/1/quiet
- Chromium61.0.3163.13/blink: does not include stroke
- Safari/WebKit: does not include stroke
- Edge25.10586.272.0: does include stroke
- Firefox54.0.1/Gecko: does include stroke

I know the code and this is easy to implement either way in both Blink and WebKit. I'm slightly leaning towards including strokes because Element.getBoundingClientRect includes CSS borders (testcase: http://output.jsbin.com/padijom/quiet). I'm slightly worried about the compatibility issues of changing Blink and WebKit.

**Comment by AmeliaBR** [comment]
I agree that something other than fill-box would be useful.  But since we've got a consistent implementation it seems strange to change it now.

At the time the CSSOM text was first drafted, there probably was only one standard definition of "bounding box" in SVG.  As far as I know, none of the browsers have yet implemented the `getBBox()` options to calculate the other bounding boxes.  Nonetheless, the CSSOM spec should probably be clarified to reference the fill bounding box, specifically.

**Comment by fsoder** [comment]
Please note that the implementations are *not* consistent (it's easy to miss the "not" in something like the above I know, I had to look twice as well...)

Let's try that again in a different form:

- Chromium/Blink: fill-box / object bounding box
- Safari/WebKit: fill-box / object bounding box
- Edge: stroke-box 
- Firefox/Gecko: stroke-box

[1]  I don't recall testing this more thoroughly, but it could be more like [2]
[2] According to testing I did a few years back I believe it was actually closer to "decorated bounding box". I don't think we have a decently matching keyword for that one though...

**Comment by AmeliaBR** [comment]
Doh. Sorry. Read too fast, first thing in the morning.

OK. So first step would be to confirm exactly what all the browsers are doing: is it stroke-box, decorated-box (including markers), or the full dirty-paint rectangle (including filter effects, like blurring or offset shadows) ?

Since the CSS layout version includes border-box and positioned children, the closest equivalent would probably be to use decorated-box: the full geometry area of the element and its descendants, but not any extra area for filters.

As an aside: the way the CSSOM definition is worded is interesting.  For elements that have both SVG and CSS layout (aka, `<svg>` elements that are root elements or children of HTML elements), it prioritizes the SVG definition over the CSS definition.  I wonder if any browsers actually do that?

**Comment by progers** [comment]
I've made two more testcases for markers and filters:
Does getBoundingClientRect include stroke? https://output.jsbin.com/fuxuqav/1/quiet
Does getBoundingClientRect include markers? https://output.jsbin.com/qaxofoj/quiet
Does getBoundingClientRect include filter offset? https://output.jsbin.com/rusakor/quiet

Firefox/Gecko and Edge do include markers, Blink and WebKit do not. No major browser includes filter offsets.

Compatibility table presented in fsoder notation:
- Chromium/Blink: object bounding box
- Safari/WebKit: object bounding box
- Edge: decorated bounding box
- Firefox/Gecko: decorated bounding box

**Comment by AmeliaBR** [comment]
Thanks for the tests Philip.

But the results I'm getting for Edge and Firefox aren't as neat and simple as "decorated bounding box" for the shapes with markers.  They're both including a lot of space beyond the dimensions added by the visible markers (but they're not consistent with each other, either).  The Firefox calculation seems to be based on the stroke-width you put on the circles in the markers (even though the context-stroke keyword isn't recognized and the markers have hidden overflow).  Which... might actually match the bounding-box algorithm, now that I think of it.   I'm really not sure what Edge is doing to calculate it's numbers.

Anyway, I'd still argue for decorated-box as being the most useful.

**Comment by progers** [comment]
I think decorated-box makes sense too, and would at least bring everyone to agreement wrt strokes.

**Comment by boggydigital** [comment]
Not blocking updated 2.0 CR publication - assigning 2.1 WD milestone

**Comment by dirkschulze** [comment]
I think the best would be to agree on a preferred behavior (obb, stroke-, or decoration box) and identify the most commonly implemented behavior. Based on that, we provide the CSS WG a proposal what it should be.

WebKit and Blink's current behavior are likely based from a time before the fork and tests probably prevented Blink to change behavior or even think about it until now.

I am confused about the different observations from @fsoder and @progers. Is Edge's implementation using object bounding box or stroke box? @AmeliaBR how is it more complicated?

**Comment by fsoder** [comment]
I've merely attempt to restate @progers observations to avoid negations. The additional notes I added were from memory at the time (which has since faded even more). So any actual deviations - if you were to see any - between mine and @progers comments should probably be accounted for by my poor memory.

As "luck" would have it, a compat bug pertaining to this was filed on Blink yesterday. The observation made there [1] did make it clear that what Gecko returns isn't quite the decorated bounding box (as already stated several times here).

[1] In this case it was a nested <svg>, on which Gecko appear to always return the viewport from gBCR (at least with overflow: hidden).

**Comment by AmeliaBR** [comment]
I have no reason to change my statement from August 2017:

> Anyway, I'd still argue for decorated-box as being the most useful.

But the complication is that the current results in Gecko and Edge aren't exactly that, and aren't consistent with each other.  So it would be spec'ing something that no one currently does. But it's more useful /consistent with CSS to include the full geometry, especially since there's already a well-supported method (`getBBox()`) to get the fill-only bounding box.
